File: src/main/scala/lox/AstPrinter.scala
package lox

final class AstPrinter extends Expr.Visitor[String] {

  override def visitAssign(assign: Expr.Assign): String =
    s"${parenthesize(s"${assign.name.lexeme} =", assign.value)}"

  override def visitBinary(binary: Expr.Binary): String =
    s"${parenthesize(binary.operator.lexeme, binary.left, binary.right)}"

  override def visitGrouping(grouping: Expr.Grouping): String =
    s"${parenthesize("group", grouping.expression)}"

  override def visitLiteral(literal: Expr.Literal): String =
    if (literal == null) "nil" else literal.value.toString

  override def visitLogical(logical: Expr.Logical): String =
    s"${parenthesize(logical.operator.lexeme, logical.left, logical.right)}"

  override def visitUnary(unary: Expr.Unary): String =
    s"${parenthesize(unary.operator.lexeme, unary.right)}"

  override def visitVariableLookup(variable: Expr.Variable): String =
    s"${parenthesize(variable.name.lexeme)}"

  private def parenthesize(name: String, expressions: Expr*): String = {
    val builder = new StringBuilder
    builder.append("(").append(name)
    for (expr <- expressions) {
      builder.append(" ")
      builder.append(expr.accept(this))
    }
    builder.append(")")
    builder.toString
  }

}

File: lox/ast_printer.go
package lox

import (
	"fmt"
	"strings"
)

type astPrinter struct{}

var AstPrinter = astPrinter{}

func (astPrinter) VisitBinaryExpr(expr *Binary) interface{} {
	return parenthesize(expr.Operator.Lexeme, expr.Left, expr.Right)
}
func (astPrinter) VisitGroupingExpr(expr *Grouping) interface{} {
	return parenthesize("group", expr.Expression)
}
func (astPrinter) VisitLiteralExpr(expr *Literal) interface{} {
	if expr.Value == nil {
		return "nil"
	}
	return fmt.Sprintf("%v", expr.Value)
}
func (astPrinter) VisitUnaryExpr(expr *Unary) interface{} {
	return parenthesize(expr.Operator.Lexeme, expr.Right)
}

func parenthesize(name string, exprs ...Expr) string {
	builder := strings.Builder{}
	builder.WriteRune('(')
	builder.WriteString(name)
	for _, expr := range exprs {
		builder.WriteRune(' ')
		builder.WriteString(expr.Accept(AstPrinter).(string))
	}
	builder.WriteRune(')')
	return builder.String()
}
